File: ios/iCam/Connectivity/Protocol/Handshake.swift
import Foundation
import CryptoKit

/// Handshake messages — `docs/PROTOCOL.md` section 4.
///
/// These are the only plaintext bytes on the connection. Both peers must
/// produce byte-identical JSON for the transcript to match, so everything here
/// goes through `HandshakeCodec`, which sorts keys and escapes nothing extra.
enum HandshakeCodec {
    static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return e
    }()
    static let decoder = JSONDecoder()

    static func encode<T: Encodable>(_ value: T) throws -> Data { try encoder.encode(value) }
    static func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        try decoder.decode(type, from: data)
    }

    /// Peeks at `t` without committing to a concrete message type.
    static func messageType(of data: Data) -> String? {
        struct Peek: Decodable { let t: String }
        return try? decoder.decode(Peek.self, from: data).t
    }
}

struct ClientHello: Codable, Sendable {
    var t = "hello"
    var v = Wire.protocolVersion
    var eph: Data          // ephemeral P-256 public key, X9.63
    var idk: Data          // identity P-256 public key, X9.63
    var rnd: Data          // 32 random bytes
    var dev: DeviceInfo
}

struct ServerHello: Codable, Sendable {
    var t = "hello_ack"
    var v = Wire.protocolVersion
    var eph: Data
    var idk: Data
    var rnd: Data
    var dev: DeviceInfo
    var sig: Data?

    /// The exact bytes the transcript covers: this message with `sig` omitted.
    func transcriptBytes() throws -> Data {
        var copy = self
        copy.sig = nil
        return try HandshakeCodec.encode(copy)
    }
}

struct ClientAuth: Codable, Sendable {
    var t = "auth"
    var sig: Data
}

struct HandshakeReady: Codable, Sendable {
    var t = "ready"
    var trusted: Bool
}

enum HandshakeError: Error, Equatable {
    case versionMismatch(Int)
    case malformed(String)
    case badSignature
    case untrusted
    case wrongMessage(expected: String, got: String)
}

/// Drives the initiator half of the handshake. The iPhone is always the
/// initiator: it browses, it connects, it proves itself first.
final class InitiatorHandshake {
    private let identity: DeviceIdentity
    private let deviceInfo: DeviceInfo
    private let ephemeral = P256.KeyAgreement.PrivateKey()
    private let clientRandom = Data((0 ..< 32).map { _ in UInt8.random(in: 0 ... 255) })

    private var helloBytes = Data()

    private(set) var peerIdentityKey = Data()
    private(set) var peerDevice: DeviceInfo?
    private(set) var keys: SessionKeys?

    init(identity: DeviceIdentity, deviceInfo: DeviceInfo) {
        self.identity = identity
        self.deviceInfo = deviceInfo
    }

    /// Step 1 — the bytes to send.
    func clientHello() throws -> Data {
        var info = deviceInfo
        info.id = identity.fingerprint
        let hello = ClientHello(eph: ephemeral.publicKey.x963Representation,
                                idk: identity.publicKeyData,
                                rnd: clientRandom,
                                dev: info)
        helloBytes = try HandshakeCodec.encode(hello)
        return helloBytes
    }

    /// Step 2 and 3 — consume the server hello, produce the client auth.
    func handle(serverHello data: Data) throws -> Data {
        let hello = try HandshakeCodec.decode(ServerHello.self, from: data)
        guard hello.t == "hello_ack" else {
            throw HandshakeError.wrongMessage(expected: "hello_ack", got: hello.t)
        }
        guard hello.v == Wire.protocolVersion else {
            throw HandshakeError.versionMismatch(hello.v)
        }
        guard let signature = hello.sig else { throw HandshakeError.malformed("missing sig") }

        var transcript = helloBytes
        transcript.append(try hello.transcriptBytes())

        var serverMessage = Data("iCam/v1/server".utf8)
        serverMessage.append(transcript)
        guard DeviceIdentity.verify(signature: signature,
                                    message: serverMessage,
                                    publicKey: hello.idk) else {
            throw HandshakeError.badSignature
        }

        guard let peerEphemeral = try? P256.KeyAgreement.PublicKey(x963Representation: hello.eph) else {
            throw HandshakeError.malformed("bad ephemeral key")
        }
        let shared = try ephemeral.sharedSecretFromKeyAgreement(with: peerEphemeral)

        keys = SessionKeys.derive(sharedSecret: shared,
                                  clientRandom: clientRandom,
                                  serverRandom: hello.rnd)
        peerIdentityKey = hello.idk
        peerDevice = hello.dev

        var clientMessage = Data("iCam/v1/client".utf8)
        clientMessage.append(transcript)
        return try HandshakeCodec.encode(ClientAuth(sig: try identity.sign(clientMessage)))
    }

    /// Step 4 — the server tells us whether it already trusts this iPhone.
    func handle(ready data: Data) throws -> Bool {
        let ready = try HandshakeCodec.decode(HandshakeReady.self, from: data)
        guard ready.t == "ready" else {
            throw HandshakeError.wrongMessage(expected: "ready", got: ready.t)
        }
        return ready.trusted
    }
}

File: windows/iCam.Core/Protocol/Handshake.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;
using System.Text.Json.Serialization;
using ICam.Core.Security;

namespace ICam.Core.Protocol;

/// <summary>
/// Canonical JSON for the handshake.
///
/// Both peers must produce byte-identical bytes or the transcript signature
/// fails, so two things are pinned here: properties are declared in sorted key
/// order (System.Text.Json writes them in declaration order, matching Swift's
/// <c>.sortedKeys</c>), and the relaxed encoder is used so a device named in
/// Arabic or Japanese is written as UTF-8 rather than as <c>\uXXXX</c> escapes,
/// which is what Swift does.
/// </summary>
public static class HandshakeCodec
{
    public static readonly JsonSerializerOptions Options = new()
    {
        Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        WriteIndented = false,
    };

    public static byte[] Encode<T>(T value) => JsonSerializer.SerializeToUtf8Bytes(value, Options);

    public static T? Decode<T>(ReadOnlySpan<byte> data) =>
        JsonSerializer.Deserialize<T>(data, Options);

    /// <summary>Peeks at <c>t</c> without committing to a concrete message type.</summary>
    public static string? MessageType(ReadOnlySpan<byte> data)
    {
        try
        {
            using var document = JsonDocument.Parse(data.ToArray());
            return document.RootElement.TryGetProperty("t", out var element)
                ? element.GetString()
                : null;
        }
        catch (JsonException)
        {
            return null;
        }
    }
}

public sealed class DeviceInfo
{
    [JsonPropertyName("app")] public string App { get; set; } = "";
    [JsonPropertyName("id")] public string Id { get; set; } = "";
    [JsonPropertyName("model")] public string Model { get; set; } = "";
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("os")] public string Os { get; set; } = "";
}

public sealed class ClientHello
{
    [JsonPropertyName("dev")] public DeviceInfo Dev { get; set; } = new();
    [JsonPropertyName("eph")] public byte[] Eph { get; set; } = [];
    [JsonPropertyName("idk")] public byte[] Idk { get; set; } = [];
    [JsonPropertyName("rnd")] public byte[] Rnd { get; set; } = [];
    [JsonPropertyName("t")] public string T { get; set; } = "hello";
    [JsonPropertyName("v")] public int V { get; set; } = Wire.ProtocolVersion;
}

public sealed class ServerHello
{
    [JsonPropertyName("dev")] public DeviceInfo Dev { get; set; } = new();
    [JsonPropertyName("eph")] public byte[] Eph { get; set; } = [];
    [JsonPropertyName("idk")] public byte[] Idk { get; set; } = [];
    [JsonPropertyName("rnd")] public byte[] Rnd { get; set; } = [];
    [JsonPropertyName("sig")] public byte[]? Sig { get; set; }
    [JsonPropertyName("t")] public string T { get; set; } = "hello_ack";
    [JsonPropertyName("v")] public int V { get; set; } = Wire.ProtocolVersion;

    /// <summary>The exact bytes the transcript covers: this message without <c>sig</c>.</summary>
    public byte[] TranscriptBytes()
    {
        var copy = new ServerHello
        {
            Dev = Dev, Eph = Eph, Idk = Idk, Rnd = Rnd, Sig = null, T = T, V = V,
        };
        return HandshakeCodec.Encode(copy);
    }
}

public sealed class ClientAuth
{
    [JsonPropertyName("sig")] public byte[] Sig { get; set; } = [];
    [JsonPropertyName("t")] public string T { get; set; } = "auth";
}

public sealed class HandshakeReady
{
    [JsonPropertyName("t")] public string T { get; set; } = "ready";
    [JsonPropertyName("trusted")] public bool Trusted { get; set; }
}

public sealed class HandshakeException : Exception
{
    public enum Reason { VersionMismatch, Malformed, BadSignature, WrongMessage }

    public Reason Kind { get; }

    public HandshakeException(Reason kind, string message) : base(message) => Kind = kind;
}

/// <summary>
/// Drives the responder half of the handshake. The PC always listens: it is the
/// stationary device, with the stable address and the firewall the user
/// controls.
/// </summary>
public sealed class ResponderHandshake : IDisposable
{
    private readonly DeviceIdentity _identity;
    private readonly DeviceInfo _deviceInfo;
    private readonly EphemeralKeyPair _ephemeral = new();
    private readonly byte[] _serverRandom = RandomNumberGenerator.GetBytes(32);

    private byte[] _transcript = [];

    public byte[] PeerIdentityKey { get; private set; } = [];
    public DeviceInfo? PeerDevice { get; private set; }
    public SessionKeys? Keys { get; private set; }

    public ResponderHandshake(DeviceIdentity identity, DeviceInfo deviceInfo)
    {
        _identity = identity;
        _deviceInfo = deviceInfo;
    }

    /// <summary>Step 1 and 2 — consume the client hello, produce the server hello.</summary>
    public byte[] HandleClientHello(ReadOnlySpan<byte> data)
    {
        var hello = HandshakeCodec.Decode<ClientHello>(data)
            ?? throw new HandshakeException(HandshakeException.Reason.Malformed,
                                            "client hello could not be read");

        if (hello.T != "hello")
        {
            throw new HandshakeException(HandshakeException.Reason.WrongMessage,
                                         $"expected hello, got {hello.T}");
        }
        if (hello.V != Wire.ProtocolVersion)
        {
            throw new HandshakeException(HandshakeException.Reason.VersionMismatch,
                                         $"peer speaks protocol version {hello.V}");
        }

        PeerIdentityKey = hello.Idk;
        PeerDevice = hello.Dev;

        var response = new ServerHello
        {
            Dev = _deviceInfo,
            Eph = _ephemeral.PublicKey,
            Idk = _identity.PublicKey,
            Rnd = _serverRandom,
        };

        // The transcript is the client's bytes exactly as they arrived, followed
        // by ours without the signature.
        var unsigned = response.TranscriptBytes();
        _transcript = new byte[data.Length + unsigned.Length];
        data.CopyTo(_transcript);
        unsigned.CopyTo(_transcript.AsSpan(data.Length));

        response.Sig = _identity.Sign(Prefixed("iCam/v1/server", _transcript));

        var shared = _ephemeral.SharedSecret(hello.Eph);
        Keys = SessionKeys.Derive(shared, hello.Rnd, _serverRandom);

        return HandshakeCodec.Encode(response);
    }

    /// <summary>Step 3 — verify the client's signature over the same transcript.</summary>
    public void HandleClientAuth(ReadOnlySpan<byte> data)
    {
        var auth = HandshakeCodec.Decode<ClientAuth>(data)
            ?? throw new HandshakeException(HandshakeException.Reason.Malformed,
                                            "client auth could not be read");
        if (auth.T != "auth")
        {
            throw new HandshakeException(HandshakeException.Reason.WrongMessage,
                                         $"expected auth, got {auth.T}");
        }
        if (!DeviceIdentity.Verify(auth.Sig, Prefixed("iCam/v1/client", _transcript),
                                   PeerIdentityKey))
        {
            throw new HandshakeException(HandshakeException.Reason.BadSignature,
                                         "the iPhone's signature did not verify");
        }
    }

    /// <summary>Step 4 — tell the iPhone whether this computer already trusts it.</summary>
    public byte[] Ready(bool trusted) =>
        HandshakeCodec.Encode(new HandshakeReady { Trusted = trusted });

    public void Dispose() => _ephemeral.Dispose();

    private static byte[] Prefixed(string label, byte[] transcript)
    {
        var prefix = Encoding.UTF8.GetBytes(label);
        var message = new byte[prefix.Length + transcript.Length];
        prefix.CopyTo(message, 0);
        transcript.CopyTo(message, prefix.Length);
        return message;
    }
}
